lib/mozart_fetcher/fetcher.ex

```diff
@@ -1,30 +1,46 @@
 defmodule MozartFetcher.Fetcher do
-  alias MozartFetcher.{Component, TimeoutParser}
+  alias MozartFetcher.{Component, TimeoutParser, Envelope}
 
   use ExMetrics
 
+  @timeout_buffer 50
+  @max_concurrency 50
+
   def process([]) do
     ExMetrics.increment("error.empty_component_list")
     Stump.log(:error, %{message: "Error cannot process empty component list"})
     {:error}
   end
 
-  def process(components) do
+  def process(components, buffer \\ @timeout_buffer) do
     ExMetrics.timeframe "function.timing.fetcher.process" do
-      max_timeout =
-        TimeoutParser.max(components) +
-          Application.get_env(:mozart_fetcher, :additional_task_await_timeout)
+      max_timeout = TimeoutParser.max(components) + buffer
+
+      stream_opts = [timeout: max_timeout,
+                     on_timeout: :kill_task,
+                     max_concurrency: @max_concurrency]
 
       components
       |> Enum.with_index()
-      |> Enum.map(&Task.async(fn -> Component.fetch(&1) end))
-      |> Enum.map(&Task.await(&1, max_timeout))
-      |> decorate_response
+      |> Task.async_stream(&Component.fetch/1, stream_opts)
+      |> validate()
+      |> decorate_response()
       |> Jason.encode!()
     end
   end
 
   defp decorate_response(envelopes) do
     %{components: envelopes}
   end
+
+  defp validate(responses) do
+    Enum.map(responses, &handle/1)
+  end
+
+  defp handle({:ok, result}), do: result
+
+  defp handle({state, reason}) do
+    Stump.log(:error, %{message: "Component Process Error", state: state, reason: reason})
+    %Envelope{}
```

Comment:
Should we fill this with a 408 or 500 status code?

Reply:
In Requester, it adds an error key to the envelope...
```go
Envelope: aggregator.Envelope{Error: err.Error()},
```

Reply:
ah, is it needed? Are you suggesting to add a fake `500` status code to the empty struct?

Reply:
Possibly. Or a 408 if we didn't get a response. I don't know if we should have error as that wasn't part of the envelope originally. We check if status is 200 so it doesn't really matter if we have it or not. The same with the error, that we don't do anything with it outside of this application.

Reply:
I think I'd prefer for now to avoid any mocked status. My only concern is that the empty envelope will not have a status key at all, this could possibly break Assembler?

Reply:
I see. I thought we had the same keys, such as id, status, index and the envelope and status were empty. I looked at this bit earlier and we check for a 200, so as long as it's safe, this will be fine.

The part that now seems an issue is the id. It looks like we raise an error if we don't get a matching id:
> https://github.com/bbc/mozart-assembler/blob/master/src/lib/mozart_assembler/validators/components.rb#L32

Reply:
We should change the logic that checks this, _nil if valid and not required, exception if not valid, or the component if successful..._